### PR TITLE
Recipe for pyopcode

### PR DESCRIPTION
**Pull request by ivoflipse**
Trying to make a recipe for our Numpy bindings for the OPCODE collision detection library

It works on Windows and Linux and at least for 3.5, hopefully also for 2.7

### CODE REVIEW

**Comment by ivoflipse** [comment]
@conda-forge/core Perhaps I shouldn't push code late in the afternoon, but what am I doing wrong here?

**Comment by jakirkham** [comment]
So, there are a few things that need changes, but I'm not sure if they are the cause of the problem or not. Let's make sure we use 2 spaces for all indentation in `meta.yaml`. Will add more comments on the diff.

File: recipes/pyopcode/meta.yml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  features:
+  - vc9                [win and py27]
+  - vc10               [win and py34]
+  - vc14               [win and py35]
+  number: 0
+
+requirements:
+  build:
+  - python
+  - numpy 1.10*
```
Comment (jakirkham) [inline recipes/pyopcode/meta.yml]: Please add `x.x`.

Is this the only version of NumPy it works with or does 1.11.x work too?

File: recipes/pyopcode/meta.yml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  - gcc 4.8.2 [linux]
+  - cmake 3.5*
+  - boost 1.6*
+  - pyyaml 3.11*
+  - setuptools
+  run:
+  - python
+  - numpy 1.10*
```
Comment (jakirkham) [inline recipes/pyopcode/meta.yml]: Also, change to `x.x`. Same question about NumPy version support.

File: recipes/pyopcode/meta.yml [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  - vc9                [win and py27]
+  - vc10               [win and py34]
+  - vc14               [win and py35]
+  number: 0
+
+requirements:
+  build:
+  - python
+  - numpy 1.10*
+  - gcc 4.8.2 [linux]
```
Comment (jakirkham) [inline recipes/pyopcode/meta.yml]: Why are we using the `gcc` package? We provide the same compiler for Linux automatically.

File: recipes/pyopcode/meta.yml [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  - cmake 3.5*
+  - boost 1.6*
+  - pyyaml 3.11*
+  - setuptools
+  run:
+  - python
+  - numpy 1.10*
+  - boost 1.6*
+  - libgcc [linux]
```
Comment (jakirkham) [inline recipes/pyopcode/meta.yml]: This should not be needed.

File: recipes/pyopcode/meta.yml [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  number: 0
+
+requirements:
+  build:
+  - python
+  - numpy 1.10*
+  - gcc 4.8.2 [linux]
+  - cmake 3.5*
+  - boost 1.6*
+  - pyyaml 3.11*
```
Comment (jakirkham) [inline recipes/pyopcode/meta.yml]: Should this be a run-time dependency?

File: recipes/pyopcode/meta.yml [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+  - vc10               [win and py34]
+  - vc14               [win and py35]
+  number: 0
+
+requirements:
+  build:
+  - python
+  - numpy 1.10*
+  - gcc 4.8.2 [linux]
+  - cmake 3.5*
```
Comment (jakirkham) [inline recipes/pyopcode/meta.yml]: Why is this pinned?

File: recipes/pyopcode/meta.yml [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+source:
+  git_url: https://github.com/ClinicalGraphics/pyopcode
+  git_tag: {{ version }}
+  
+build:
+  features:
+  - vc9                [win and py27]
+  - vc10               [win and py34]
+  - vc14               [win and py35]
```
Comment (jakirkham) [inline recipes/pyopcode/meta.yml]: All selectors should be like `# [win and py35]`.

File: recipes/pyopcode/meta.yml [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+ test:
+  commands:
+  - py.test --pyargs pyopcode
+  imports:
+  - pyopcode
+  requires:
+  - pytest 2.9*
+  
+about:
+  home: hhttps://github.com/ClinicalGraphics/pyopcode
```
Comment (jakirkham) [inline recipes/pyopcode/meta.yml]: `hhttps` -> `https`

File: recipes/pyopcode/meta.yml [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+  license: LGPL
```
Comment (jakirkham) [inline recipes/pyopcode/meta.yml]: Please add the version of LGPL.

File: recipes/pyopcode/meta.yml [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+  summary: OPCODE collision detection wrapped library with numpy bindings
+  
+extra:
+  recipe-maintainers:
+    - ivoflipse
+    - Korijn
+    - Maxyme
+    - EelcoHoogendoorn
```
Comment (jakirkham) [inline recipes/pyopcode/meta.yml]: Please add a terminal newline.

File: recipes/pyopcode/meta.yml [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+  - numpy 1.10*
+  - gcc 4.8.2 [linux]
+  - cmake 3.5*
+  - boost 1.6*
+  - pyyaml 3.11*
+  - setuptools
+  run:
+  - python
+  - numpy 1.10*
+  - boost 1.6*
```
Comment (jakirkham) [inline recipes/pyopcode/meta.yml]: Should be 1.61.*.

**Comment by jakirkham** [comment]
Pinnings can be found [here]( https://github.com/conda-forge/conda-forge.github.io/blob/71e87bb41a42da65e6b63446e974ed57d536a0b7/scripts/pin_the_slow_way.py#L32-L48 ) now.

File: recipes/pyopcode/bld.bat [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+if %PY_VER%==2.7 (
+    set GENERATOR_NAME=Visual Studio 9 2008
+)
+if %PY_VER%==3.5 (
+    set GENERATOR_NAME=Visual Studio 14 2015
+)
+
+REM pick architecture
+if %ARCH%==64 (
+	set GENERATOR_NAME=%GENERATOR_NAME% Win64
```
Comment (jakirkham) [inline recipes/pyopcode/bld.bat]: Would be worth trying `%CMAKE_GENERATOR%` as it is set by `conda-build` to handle this stuff. See [here]( http://conda.pydata.org/docs/building/environment-vars.html ).

File: recipes/pyopcode/meta.yml (hunk 1 above)
Comment (ivoflipse) [inline recipes/pyopcode/meta.yml]: Not sure what versions it will work for. I don't think it uses any fancy numpy features

File: recipes/pyopcode/meta.yaml [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+  - numpy x.x
+  - boost 1.61.*
+
+test:
+  commands:
+    - py.test --pyargs pyopcode
+  imports:
+    - pyopcode
+  requires:
+    - pytest 2.9*
```
Comment (Korijn) [inline recipes/pyopcode/meta.yaml]: I think this version also doesn't need to be pinned

**Comment by ivoflipse** [comment]
@jakirkham I'm also getting struck by the VS 2008 x64 problem it seems. Anything I can do from my end?

File: recipes/pyopcode/build.sh [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+    if [ ! -f $PYTHON_LIBRARY ]; then
+        PYTHON_LIBRARY=${PREFIX}/lib/libpython${PY_VER}m.so
+    fi
+
+    cd pyopcode
+    mkdir build
+    cd build
+
+    cmake ../src \
+        -Wno-dev \
```
Comment (jakirkham) [inline recipes/pyopcode/build.sh]: I think this flag might be [confusing]( https://ci.appveyor.com/project/conda-forge/staged-recipes/build/1.0.3328/job/9fmtgpv3qibeakgg#L330 ) it.

File: recipes/pyopcode/build.sh [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+    PYTHON_LIBRARY=${PREFIX}/lib/${PY_LIB}
+    if [ ! -f $PYTHON_LIBRARY ]; then
+        PYTHON_LIBRARY=${PREFIX}/lib/libpython${PY_VER}m.so
+    fi
+
+    cd pyopcode
+    mkdir build
+    cd build
+
+    cmake ../src \
```
Comment (Korijn) [inline recipes/pyopcode/build.sh]: Actually, I think you're missing the generator here

File: recipes/pyopcode/build.sh (hunk 14 above)
Comment (Korijn) [inline recipes/pyopcode/build.sh]: The error says that it can't recognize generator `-Wno-dev`, since cmake expects the first argument to be `-G{generator_name_here}` and its left out

File: recipes/pyopcode/bld.bat [hunk 16]
```diff
@@ -0,0 +1,29 @@
+set BUILD_CONFIG=Release
+REM tell cmake where Python is
+set PYTHON_LIBRARY=%PREFIX%\libs\python%PY_VER:~0,1%%PY_VER:~2,1%.lib
+
+REM work in build subdir
+cd pyopcode
+mkdir build
+cd build
+
+cmake ../src -G%CMAKE_GENERATOR% ^
```
Comment (Korijn) [inline recipes/pyopcode/bld.bat]: %CMAKE_GENERATOR% appears to be empty

File: recipes/pyopcode/bld.bat [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+    -Wno-dev ^
```
Comment (jakirkham) [inline recipes/pyopcode/bld.bat]: I think this flag might be [confusing]( https://ci.appveyor.com/project/conda-forge/staged-recipes/build/1.0.3328/job/9fmtgpv3qibeakgg#L330 ) it.

File: recipes/pyopcode/build.sh (hunk 14 above)
Comment (jakirkham) [inline recipes/pyopcode/build.sh]: Ack, sorry, I meant to put that comment in the `bld.bat`. Have now added it up there.

File: recipes/pyopcode/bld.bat (hunk 17 above)
Comment (ivoflipse) [inline recipes/pyopcode/bld.bat]: Strange, because a colleague used a version of this bld.bat file to build it for 2.7 where the flag was there. I've reverted back to using a manually constructed GENERATOR_NAME to see if that does the trick

**Comment by ivoflipse** [comment]
My last change fixed 3.4, broke 3.5 and left 2.7 broken...

**Comment by ivoflipse** [comment]
Switching to using a tarball prevents the cloning issue, so now only Python 2.7 for x64 is a problem

**Comment by patricksnape** [comment]
I recommend nmake on Windows, rather than all this logic for the correct visual studio generator.

**Comment by ivoflipse** [comment]
Any concrete examples that I can look at to switch it to nmake @patricksnape?

**Comment by patricksnape** [comment]
https://github.com/patricksnape/opencv-feedstock/blob/8d1049690ed18a9a05699463989c7a133a11858d/recipe/bld.bat

**Comment by ivoflipse** [comment]
Two steps forward, one step back...
```
C:\Miniconda-x64\conda-bld\work\pyopcode-0.2.0\pyopcode>cmake --build ./build --clean-first --target ALL_BUILD --config Release 
NMAKE : fatal error U1073: don't know how to make 'ALL_BUILD'
```

**Comment by ivoflipse** [comment]
@patricksnape I reckon switching to NMake also requires me to make changes to the CMakeLists.txt file? Since I changed the script to be pretty much identical to your OpenCV script, but I reckon you aren't getting these errors ;-)

File: recipes/pyopcode/bld.bat [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+    -DCMAKE_BUILD_TYPE=%BUILD_CONFIG% ^
+    -DCMAKE_INSTALL_PREFIX="%PREFIX%" ^
+    -DPYTHON_INCLUDE_DIR:PATH="%PREFIX%/include" ^
+    -DPYTHON_LIBRARY:FILEPATH="%PYTHON_LIBRARY%" ^
+    -DNUMPY_INCLUDE_DIR:PATH="%SP_DIR%/numpy/core/include" ^
+    -DBOOST_ROOT:PATH="%PREFIX%/Library"
+if errorlevel 1 exit 1
+
+echo dir /p
+cmake --build . --target INSTALL --config %BUILD_CONFIG%
```
Comment (patricksnape) [inline recipes/pyopcode/bld.bat]: Try just calling `nmake` and then copying the built file.

**Comment by ivoflipse** [comment]
W00t! Thanks @patricksnape! It's working now.

@jakirkham now this PR looks good to me

File: recipes/pyopcode/build.sh [hunk 19]
```diff
@@ -0,0 +1,46 @@
+#!/bin/bash
+BUILD_CONFIG=Release
+
+if [ `uname` == Linux ]; then
+    CC=${PREFIX}/bin/gcc
+    CXX=${PREFIX}/bin/g++
```
Comment (jakirkham) [inline recipes/pyopcode/build.sh]: I'm guessing these are holdovers that we can clear.

File: recipes/pyopcode/meta.yaml [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+  - setuptools
+  - pyyaml
+  run:
+  - python
+  - numpy x.x
+  - boost 1.61.*
+
+test:
+  commands:
+    - py.test --pyargs pyopcode
```
Comment (jakirkham) [inline recipes/pyopcode/meta.yaml]: 👍

File: recipes/pyopcode/build.sh [hunk 21]
```diff
@@ -0,0 +1,46 @@
+#!/bin/bash
+BUILD_CONFIG=Release
+
```
Comment (jakirkham) [inline recipes/pyopcode/build.sh]: Could we try using the `toolchain`? Please see this PR ( https://github.com/conda-forge/thrift-cpp-feedstock/pull/4 ) for and example of how to do this.

File: recipes/pyopcode/build.sh [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+    if [ ! -d $include_path ]; then
+      # Control will enter here if $DIRECTORY doesn't exist.
+      include_path=${PREFIX}/include/python${PY_VER}m
+    fi
+
+    PY_LIB="libpython${PY_VER}.so"
+    PYTHON_LIBRARY=${PREFIX}/lib/${PY_LIB}
+    if [ ! -f $PYTHON_LIBRARY ]; then
+        PYTHON_LIBRARY=${PREFIX}/lib/libpython${PY_VER}m.so
+    fi
```
Comment (jakirkham) [inline recipes/pyopcode/build.sh]: Yeah, this is fun. I'm ok with leaving it in here for now. Though I would recommend trying to get this into CMake somehow.

I tried to use CMake's "FindPython*" and got frustrated. So, I wrote [my own]( https://github.com/nanshe-org/rank_filter/blob/94b783fbaf20c853b1bec1b64ecfdbaea5d7fb1a/cmake/modules/FindPYTHON.cmake ). Feel free to take it if it helps. Basically it uses the Python interpreter to find out all of this information. Though I haven't tried it on Windows; so, I don't know if it works there or not.

**Comment by EelcoHoogendoorn** [comment]
There are a bunch of things that could still be improved here i think. First of all, the version pinning is overly specific; 1.61 for boost, for instance. Only the most vanilla boost python features are used, so any version would do (i imagine...). The reason for pinning at 1.61 was to force the use of conda-forge boost; which unlike the one on the defaults channel, does contain runtime libs.

Which brings me to another point; I dont think this package uses the runtime libs in any way, so boost shouldnt need to be a runtime dependency; boost is only used for header includes. My boost cmake related stuff is simply a copy-paste from something i found elsewhere that worked; but simply stripping out the linking of the libraries does not fly. But i bet it should very well be possible to build this without a runtime dependency on boost, for someone who isnt a complete cmake noob such as myself.

**Comment by ivoflipse** [comment]
I removed the leftovers regarding setting gcc and such, but the other changes require changes to the CMAKELIST of pyopcode correct? So I'd propose to merge this PR and then start playing around with it on the feedstock itself

**Comment by jakirkham** [comment]
> There are a bunch of things that could still be improved here i think. First of all, the version pinning is overly specific; 1.61 for boost, for instance. Only the most vanilla boost python features are used, so any version would do (i imagine...). The reason for pinning at 1.61 was to force the use of conda-forge boost; which unlike the one on the defaults channel, does contain runtime libs.

We have pinned `boost` tightly on a [global scale]( https://github.com/conda-forge/conda-forge.github.io/blob/7d6a2022906110b042bb3ba90a1c27695a692302/scripts/pin_the_slow_way.py#L32 ). So, it is not specific to this recipe perse, @EelcoHoogendoorn. As we have seen [many ABI breaks]( http://upstream.rosalinux.ru/versions/boost.html ) in Boost during minor version bumps, we have decided to be very cautious and pin the major and minor numbers of Boost.

However, we do not have a recent ABI report. So, it makes it tricky to decide when this could be skipped or not. Also, we cannot really predict future breaks so doing something like `1.6*` is an unsafe choice. In practice, the degree of pinning used here ends up being safe.

I should add that we have the ability to send pinning PRs through the stack when we do upgrade Boost. So, in practice, this tight pinning is not so debilitating as one might assume.

> Which brings me to another point; I dont think this package uses the runtime libs in any way, so boost shouldnt need to be a runtime dependency; boost is only used for header includes. 

So, unfortunately, this is just not the case if one uses Boost.Python. Instead, one will be linked to dynamically linked to `libboost_python`. As a result, we do need to pass that library around as a run-time dependency.

> My boost cmake related stuff is simply a copy-paste from something i found elsewhere that worked; but simply stripping out the linking of the libraries does not fly. But i bet it should very well be possible to build this without a runtime dependency on boost, for someone who isnt a complete cmake noob such as myself.

If one went the static linking route, I suppose it might be possible. However, static linking is generally frowned upon at conda-forge. One must have some extraordinary circumstances for that to be a good idea.

<br>
In short, I think we are doing the right thing here w.r.t. to Boost.

**Comment by EelcoHoogendoorn** [comment]
Ah great, thanks for the clarification!

**Comment by ivoflipse** [comment]
@jakirkham Can this be merged now?
